File: water_demo.py
#!/usr/bin/env python3
"""
Water Body Detection Demo - Python Implementation
Demonstrates the same NDWI functionality as the Rust implementation
"""

import numpy as np
import matplotlib.pyplot as plt
from datetime import datetime
import json
import os

class WaterMonitor:
    def __init__(self, threshold=0.3, alert_threshold_pct=10.0):
        self.threshold = threshold
        self.alert_threshold_pct = alert_threshold_pct
    
    def compute_ndwi(self, green, nir):
        """Compute NDWI: (Green - NIR) / (Green + NIR)"""
        with np.errstate(divide='ignore', invalid='ignore'):
            ndwi = (green - nir) / (green + nir)
            ndwi = np.where(np.isfinite(ndwi), ndwi, 0.0)
        return ndwi
    
    def threshold_ndwi(self, ndwi):
        """Create binary water mask"""
        return (ndwi > self.threshold).astype(np.uint8)
    
    def generate_mock_satellite_data(self, width=100, height=100):
        """Generate mock satellite data with water bodies"""
        green = np.zeros((height, width))
        nir = np.zeros((height, width))
        
        # Create coordinate grids
        y, x = np.ogrid[:height, :width]
        center_x, center_y = width // 2, height // 2
        
        # Main lake in center
        main_lake = (x - center_x)**2 + (y - center_y)**2 < 15**2
        
        # Smaller ponds
        pond1 = (x - 20)**2 + (y - 20)**2 < 8**2
        pond2 = (x - 80)**2 + (y - 60)**2 < 6**2
        
        # Water areas
        water_mask = main_lake | pond1 | pond2
        
        # Set reflectance values
        green[water_mask] = 0.8 + np.random.normal(0, 0.05, np.sum(water_mask))
        green[~water_mask] = 0.2 + np.random.normal(0, 0.05, np.sum(~water_mask))
        
        nir[water_mask] = 0.1 + np.random.normal(0, 0.05, np.sum(water_mask))
        nir[~water_mask] = 0.6 + np.random.normal(0, 0.1, np.sum(~water_mask))
        
        # Ensure values are in valid range
        green = np.clip(green, 0, 1)
        nir = np.clip(nir, 0, 1)
        
        return green, nir
    
    def detect_water_bodies(self, green, nir, output_dir="./water_output"):
        """Process satellite data for water body detection"""
        print("🛰️  Detecting water bodies from satellite data")
        
        # Create output directory
        os.makedirs(output_dir, exist_ok=True)
        
        # 1. Compute NDWI
        ndwi = self.compute_ndwi(green, nir)
        
        # 2. Threshold to get water mask
        water_mask = self.threshold_ndwi(ndwi)
        
        # 3. Calculate statistics
        total_pixels = water_mask.size
        water_pixels = np.sum(water_mask)
        water_percentage = (water_pixels / total_pixels) * 100
        
        # 4. Estimate water area (30m Landsat pixels)
        pixel_area = 900.0  # m²
        total_area = water_pixels * pixel_area
        
        # 5. Save visualizations
        self.save_results(green, nir, ndwi, water_mask, output_dir)
        
        # 6. Save statistics
        stats = {
            "timestamp": datetime.now().isoformat(),
            "total_pixels": int(total_pixels),
            "water_pixels": int(water_pixels),
            "water_percentage": float(water_percentage),
            "total_area_m2": float(total_area),
            "ndwi_threshold": float(self.threshold),
            "ndwi_min": float(np.min(ndwi)),
            "ndwi_max": float(np.max(ndwi)),
            "ndwi_mean": float(np.mean(ndwi))
        }
        
        with open(f"{output_dir}/water_stats.json", "w") as f:
            json.dump(stats, f, indent=2)
        
        print(f"💧 Detected {water_pixels} water pixels ({water_percentage:.1f}%) = {total_area:.0f} m²")
        return total_area
    
    def save_results(self, green, nir, ndwi, water_mask, output_dir):
        """Save visualization results"""
        fig, axes = plt.subplots(2, 2, figsize=(12, 10))
        fig.suptitle('Water Body Detection Results', fontsize=16)
        
        # Green band
        im1 = axes[0,0].imshow(green, cmap='Greens')
        axes[0,0].set_title('Green Band')
        axes[0,0].axis('off')
        plt.colorbar(im1, ax=axes[0,0])
        
        # NIR band
        im2 = axes[0,1].imshow(nir, cmap='Reds')
        axes[0,1].set_title('NIR Band')
        axes[0,1].axis('off')
        plt.colorbar(im2, ax=axes[0,1])
        
        # NDWI
        im3 = axes[1,0].imshow(ndwi, cmap='RdYlBu', vmin=-1, vmax=1)
        axes[1,0].set_title('NDWI')
        axes[1,0].axis('off')
        plt.colorbar(im3, ax=axes[1,0])
        
        # Water mask
        im4 = axes[1,1].imshow(water_mask, cmap='Blues')
        axes[1,1].set_title('Water Mask')
        axes[1,1].axis('off')
        plt.colorbar(im4, ax=axes[1,1])
        
        plt.tight_layout()
        plt.savefig(f"{output_dir}/water_detection_results.png", dpi=150, bbox_inches='tight')
        plt.close()
        
        print(f"📁 Results saved to: {output_dir}/")
    
    def check_for_alerts(self, aoi_id, prev_area, curr_area, next_rain_days=None):
        """Check for drought alerts"""
        if prev_area <= 0:
            return None
        
        drop_pct = (prev_area - curr_area) / prev_area * 100
        
        if drop_pct > self.alert_threshold_pct:
            if drop_pct > 30:
                level = "CRITICAL"
                emoji = "🚨"
            elif drop_pct > 20:
                level = "WARNING"
                emoji = "⚠️"
            else:
                level = "INFO"
                emoji = "ℹ️"
            
            alert = {
                "aoi_id": aoi_id,
                "prev_area": prev_area,
                "curr_area": curr_area,
                "drop_pct": drop_pct,
                "timestamp": datetime.now().isoformat(),
                "next_rain_days": next_rain_days,
                "alert_level": level
            }
            
            area_drop = prev_area - curr_area
            rain_msg = f"Next rain in {next_rain_days} days." if next_rain_days else "No rain forecast."
            
            message = f"{emoji} {level} Lake area dropped by {area_drop:.0f} m² ({drop_pct:.1f}%) this week. {rain_msg} Watch for drought risk."
            
            return alert, message
        
        return None

def main():
    print("🚀 Starting Water Body Detection Demo")
    
    # Initialize monitor
    monitor = WaterMonitor()
    
    # Generate mock satellite data
    print("📊 Generating mock satellite data (100x100 pixels)...")
    green, nir = monitor.generate_mock_satellite_data(100, 100)
    
    # Process the data
    current_area = monitor.detect_water_bodies(green, nir)
    
    # Simulate temporal analysis
    previous_area = 45000.0  # Previous measurement
    
    print("\n📈 Temporal Analysis:")
    print(f"Previous water area: {previous_area:.0f} m²")
    print(f"Current water area:  {current_area:.0f} m²")
    
    # Check for alerts
    result = monitor.check_for_alerts("lake-001", previous_area, current_area, 5)
    
    if result:
        alert, message = result
        print(f"\n{message}")
        print(f"Alert details: {json.dumps(alert, indent=2)}")
    else:
        print("\n✅ Water area stable - no alerts triggered")
    
    print("\n🎉 Water body detection demo completed successfully!")

if __name__ == "__main__":
    main()

File: water_demo.rs
use anyhow::Result;
use std::path::Path;
use chrono::Utc;

// Standalone demo without shared crate dependencies

#[derive(Debug, Clone)]
pub struct BBox {
    pub min_lon: f64,
    pub min_lat: f64,
    pub max_lon: f64,
    pub max_lat: f64,
}

#[derive(Debug, Clone)]
pub struct AOI {
    pub id: String,
    pub bbox: BBox,
    pub name: Option<String>,
}

#[derive(Debug, Clone)]
pub enum AlertLevel {
    Info,
    Warning,
    Critical,
}

#[derive(Debug, Clone)]
pub struct WaterAlert {
    pub aoi_id: String,
    pub prev_area: f64,
    pub curr_area: f64,
    pub drop_pct: f64,
    pub timestamp: chrono::DateTime<chrono::Utc>,
    pub next_rain_days: Option<u32>,
    pub alert_level: AlertLevel,
}

/// Water Body Monitoring Service
pub struct WaterMonitor {
    threshold: f32,
    alert_threshold_pct: f64,
}

impl WaterMonitor {
    pub fn new() -> Self {
        Self {
            threshold: 0.3,
            alert_threshold_pct: 10.0,
        }
    }

    /// Compute NDWI from band data arrays
    pub fn compute_ndwi_from_arrays(&self, green: &[f32], nir: &[f32]) -> Result<Vec<f32>> {
        if green.len() != nir.len() {
            anyhow::bail!("Green and NIR arrays must have the same length");
        }
        
        let ndwi: Vec<f32> = green.iter().zip(nir.iter())
            .map(|(&g, &n)| if (g + n).abs() > 1e-6 { (g - n) / (g + n) } else { 0.0 })
            .collect();
        
        Ok(ndwi)
    }

    /// Threshold NDWI to create a binary water mask
    pub fn threshold_ndwi(&self, ndwi: &[f32]) -> Vec<u8> {
        ndwi.iter().map(|&v| if v > self.threshold { 1 } else { 0 }).collect()
    }

    /// Process band arrays for water body detection
    pub async fn detect_water_bodies_from_arrays(
        &self, 
        green: &[f32], 
        nir: &[f32], 
        width: usize, 
        height: usize,
        output_dir: &Path
    ) -> Result<f64> {
        println!("🛰️  Detecting water bodies from {}x{} satellite data", width, height);

        // Create output directory
        std::fs::create_dir_all(output_dir)?;

        // 1. Compute NDWI
        let ndwi = self.compute_ndwi_from_arrays(green, nir)?;
        
        // 2. Threshold NDWI to get water mask
        let mask = self.threshold_ndwi(&ndwi);
        
        // 3. Calculate water statistics
        let water_pixels = mask.iter().filter(|&&v| v == 1).count();
        let total_pixels = mask.len();
        let water_percentage = (water_pixels as f64 / total_pixels as f64) * 100.0;
        
        // 4. Estimate water area (assuming 30m Landsat pixels)
        let pixel_area = 900.0; // 30m x 30m
        let total_area = water_pixels as f64 * pixel_area;

        // 5. Save results
        let ndwi_stats = format!(
            "NDWI Statistics:\n\
             - Total pixels: {}\n\
             - Water pixels: {} ({:.1}%)\n\
             - Estimated water area: {:.0} m²\n\
             - NDWI threshold: {:.2}",
            total_pixels, water_pixels, water_percentage, total_area, self.threshold
        );
        
        let results_path = output_dir.join("water_detection_results.txt");
        std::fs::write(results_path, ndwi_stats)?;

        println!("💧 Detected {} water pixels ({:.1}%) = {:.0} m²", 
                 water_pixels, water_percentage, total_area);
        
        Ok(total_area)
    }

    /// Monitor water body changes and generate alerts
    pub fn check_for_alerts(&self, aoi: &AOI, prev_area: f64, curr_area: f64, next_rain_days: Option<u32>) -> Option<WaterAlert> {
        if prev_area <= 0.0 {
            return None;
        }

        let drop_pct = (prev_area - curr_area) / prev_area * 100.0;
        
        if drop_pct > self.alert_threshold_pct {
            let alert_level = if drop_pct > 30.0 {
                AlertLevel::Critical
            } else if drop_pct > 20.0 {
                AlertLevel::Warning
            } else {
                AlertLevel::Info
            };

            Some(WaterAlert {
                aoi_id: aoi.id.clone(),
                prev_area,
                curr_area,
                drop_pct,
                timestamp: Utc::now(),
                next_rain_days,
                alert_level,
            })
        } else {
            None
        }
    }

    /// Format alert message for notification
    pub fn format_alert_message(&self, alert: &WaterAlert) -> String {
        let area_drop = alert.prev_area - alert.curr_area;
        let rain_msg = if let Some(days) = alert.next_rain_days {
            format!("Next rain in {} days.", days)
        } else {
            "No rain forecast available.".to_string()
        };

        let severity = match alert.alert_level {
            AlertLevel::Critical => "🚨 CRITICAL",
            AlertLevel::Warning => "⚠️ WARNING",
            AlertLevel::Info => "ℹ️ INFO",
        };

        format!(
            "{} Lake area dropped by {:.0} m² ({:.1}%) this week. {} Watch for drought risk.",
            severity, area_drop, alert.drop_pct, rain_msg
        )
    }

    /// Generate mock satellite data for demonstration
    pub fn generate_mock_satellite_data(width: usize, height: usize) -> (Vec<f32>, Vec<f32>) {
        let mut green = Vec::with_capacity(width * height);
        let mut nir = Vec::with_capacity(width * height);
        
        for y in 0..height {
            for x in 0..width {
                // Create a pattern with multiple "water bodies"
                let center_x = width / 2;
                let center_y = height / 2;
                let distance = ((x as f32 - center_x as f32).powi(2) + (y as f32 - center_y as f32).powi(2)).sqrt();
                
                // Create a "lake" in center
                let in_main_lake = distance < 15.0;
                
                // Create smaller "ponds"
                let in_pond1 = ((x as f32 - 20.0).powi(2) + (y as f32 - 20.0).powi(2)).sqrt() < 8.0;
                let in_pond2 = ((x as f32 - 80.0).powi(2) + (y as f32 - 60.0).powi(2)).sqrt() < 6.0;
                
                if in_main_lake || in_pond1 || in_pond2 {
                    // Water area - higher green reflectance, lower NIR
                    green.push(0.8 + (x as f32 / width as f32) * 0.1); // Slight variation
                    nir.push(0.1 + (y as f32 / height as f32) * 0.1);
                } else {
                    // Land/vegetation - lower green, higher NIR
                    green.push(0.2 + (x as f32 / width as f32) * 0.2);
                    nir.push(0.6 + (y as f32 / height as f32) * 0.2);
                }
            }
        }
        
        (green, nir)
    }
}

#[tokio::main]
async fn main() -> Result<()> {
    println!("🚀 Starting Water Body Detection Demo");
    
    // Example AOI
    let aoi = AOI {
        id: "lake-001".to_string(),
        bbox: BBox {
            min_lon: -120.0,
            min_lat: 35.0,
            max_lon: -119.0,
            max_lat: 36.0,
        },
        name: Some("Demo Lake".to_string()),
    };

    let monitor = WaterMonitor::new();
    let output_dir = Path::new("./water_output");
    
    println!("📊 Generating mock satellite data (100x100 pixels)...");
    
    // Generate mock satellite data
    let (green, nir) = monitor.generate_mock_satellite_data(100, 100);
    
    // Process the mock data
    let current_area = monitor.detect_water_bodies_from_arrays(&green, &nir, 100, 100, output_dir).await?;
    
    // Simulate previous measurement for comparison
    let previous_area = 45000.0; // Previous area in m²
    
    println!("\n📈 Temporal Analysis:");
    println!("Previous water area: {:.0} m²", previous_area);
    println!("Current water area:  {:.0} m²", current_area);
    
    // Check for alerts
    if let Some(alert) = monitor.check_for_alerts(&aoi, previous_area, current_area, Some(5)) {
        let message = monitor.format_alert_message(&alert);
        println!("\n{}", message);
        println!("Alert details: {:?}", alert);
    } else {
        println!("\n✅ Water area stable - no alerts triggered");
    }

    println!("\n📁 Results saved to: {}", output_dir.display());
    println!("🎉 Water body detection demo completed successfully!");

    Ok(())
}
